include/i2c.h

```diff
@@ -110,16 +110,63 @@ union __deprecated dev_config {
  * These are for internal use only, so skip these in
  * public documentation.
  */
+struct i2c_slave_callbacks;
+
+typedef int (*i2c_slave_write_requested_cb_t)(
+		struct i2c_slave_callbacks *callbacks);
+typedef int (*i2c_slave_read_requested_cb_t)(
+		struct i2c_slave_callbacks *callbacks, u8_t *val);
+typedef int (*i2c_slave_write_received_cb_t)(
+		struct i2c_slave_callbacks *callbacks, u8_t val);
+typedef int (*i2c_slave_read_processed_cb_t)(
+		struct i2c_slave_callbacks *callbacks, u8_t *val);
+typedef int (*i2c_slave_stop_cb_t)(struct i2c_slave_callbacks *callbacks);
+
+struct i2c_slave_callbacks {
+	/** This is meant to be used in the driver and the user should not
+	 * mess with it
+	 */
+	sys_snode_t node;
+	/** flags for this slave device */
+	u8_t flags;
+	/** address for this slave device */
+	u16_t address;
+	/** callback function being called when write is requested */
+	i2c_slave_write_requested_cb_t write_requested;
```

Comment:
Ok

Reply:
Instead of having 5 callbacks which is a bit heavy, what about 1 taking a code as parameters (which code would follow an enum for all these events)?

Reply:
I see some advantages on using 5 callbacks instead of a single callback with a switch case:
- multiple callbacks only take 5 pointers to define then and avoid all the switch/case/parameters code
- in a switch/case handling code, the compiler won't be able to suppress unused cases, you will always compile and ship the entire function, but if you split them you will have the exact same code size (minus the parameter and switch/case comparisong) but the compiler can drop unused functions
- is this a blocker ?

Reply:
In slave side it's only a matter of 5 case, and if the dev is doing it right each case would call an inline function for each case.

The compiler is not going to drop unused callbacks since they are hooked to the i2c device at runtime through registration, and these pointer have to be always valid also according to your stm32 implementation.
There is nothing to suppress at compilation time in fact.

You won't have the same code size no, properly handled the callback could be called only 2 times (according to your stm32 implementation).

Reply:
Sorry @tbursztyka but I don't understand the point here, it is a showstopper ? does it break some coding/API rules ? Can we go further at some point ?
@DWagenk did a great job adding multi-master to make sure the API is correct, what should we do now ?

Reply:
I was just raising the fact it is a bit bloated. Zephyr is meant to run on memory constrained SoC after all. On an OS such as Linux I totally understand that would be the thing to do.
That why, usually, when designing an API we should try to address being generic as well as being memory friendly.

Reply:
OK, thanks for your fact, I do not consider having a single handler is memory friendlier, except the only fact that the  i2c_slave_callbacks will use 5 words of storage instead of 1, but the 5 comparison in the single handler will take at least 2 instructions per event case.
But if you personally consider it's not beautiful enough, it's another issue.
For now I don't consider it's worth spending 2hours of my time rebasing and fixing the callbacks, considering this is correct and working code.